### PR TITLE
revise-code-for-catchability-with-bias-correction

### PR DESCRIPTION
**Pull request by Rick-Methot-NOAA**
* Resolves issue #535

## What tests have been done? 
<-- - [] Test files are in the issue. -->

### What tests/review still need to be done?
needs close look by a math-stat whiz

## Is there an input change for users to Stock Synthesis? 
revision for T-dist

### CODE REVIEW

File: SS_objfunc.tpl [hunk 1]
```diff
@@ -96,69 +97,56 @@ FUNCTION void evaluate_the_objective_function()
...
           {
-            if (Svy_use(f, i) > 0)
-            {
-              surv_like(f) += 0.5 * square((Svy_obs(f, i) - Svy_est(f, i)) / Svy_se_use(f, i)) + sd_offset * log(Svy_se_use(f, i));
+            if (Svy_errtype(f) == 0) {  // lognormal
+              Svy_like_I(f,i) = 0.5 * square((Svy_obs_log(f, i) - Svy_est(f, i)) / Svy_se_use(f, i))
+                              + sd_offset * log(Svy_se_use(f, i));
+             }
+            else if (Svy_errtype(f) == 1) {  // lognormal with bias adjustment
+              Svy_like_I(f,i) = 0.5 * square((Svy_obs_log(f, i) - Svy_est(f, i) + 0.5 * Svy_se_use(f, i) * Svy_se_use(f, i) ) / Svy_se_use(f, i))
```
Comment (Cole-Monnahan-NOAA) [inline SS_objfunc.tpl:136]: Why not use `square(Svy_se_use(f,i))`?

File: SS_objfunc.tpl [hunk 2]
```diff
@@ -96,69 +97,56 @@ FUNCTION void evaluate_the_objective_function()
...
-          for (i = 1; i <= Svy_N_fleet(f); i++)
+        for (i = 1; i <= Svy_N_fleet(f); i++)
+          if (Svy_use(f, i) > 0)
           {
-            if (Svy_use(f, i) > 0)
-            {
-              surv_like(f) += 0.5 * square((Svy_obs(f, i) - Svy_est(f, i)) / Svy_se_use(f, i)) + sd_offset * log(Svy_se_use(f, i));
+            if (Svy_errtype(f) == 0) {  // lognormal
+              Svy_like_I(f,i) = 0.5 * square((Svy_obs_log(f, i) - Svy_est(f, i)) / Svy_se_use(f, i))
+                              + sd_offset * log(Svy_se_use(f, i));
```
Comment (Cole-Monnahan-NOAA) [inline SS_objfunc.tpl:133]: I don't follow this sd_offset component.

File: SS_objfunc.tpl [hunk 3]
```diff
@@ -96,69 +97,56 @@ FUNCTION void evaluate_the_objective_function()
...
+             }
+            else if (Svy_errtype(f) == 1) {  // lognormal with bias adjustment
+              Svy_like_I(f,i) = 0.5 * square((Svy_obs_log(f, i) - Svy_est(f, i) + 0.5 * Svy_se_use(f, i) * Svy_se_use(f, i) ) / Svy_se_use(f, i))
+                              + sd_offset * log(Svy_se_use(f, i));
+              }
+            else if (Svy_errtype(f) >= 1000) {  // T-dist
```
Comment (Cole-Monnahan-NOAA) [inline SS_objfunc.tpl:139]: Shouldn't this be >1? What if someone used a t-dist with df=4 before. Won't this break it?

File: SS_objfunc.tpl [hunk 4]
```diff
@@ -96,69 +97,56 @@ FUNCTION void evaluate_the_objective_function()
...
+              dvariable df = Svy_errtype(f) - 1000.;
+              Svy_like_I(f,i) = ((df + 1.) / 2.) * log((1. + square((Svy_obs_log(f, i) - Svy_est(f, i))) / (df * square(Svy_se_use(f, i)))))
+                              + sd_offset * log(Svy_se_use(f, i));
+              }
+            else if (Svy_errtype(f) == -1) {  // normal
+              Svy_like_I(f,i) =  0.5 * square((Svy_obs(f, i) - Svy_est(f, i)) / Svy_se_use(f, i))
+                              + sd_offset * log(Svy_se_use(f, i));
             }
           }
```
Comment (Cole-Monnahan-NOAA) [inline SS_objfunc.tpl:148]: Maybe this is caught earlier but what if an invalid Svy_errtype is entered? Say a -5? It looks like it'll just skip the calculations but really should error out or warn?

File: SS_readdata_330.tpl [hunk 5]
```diff
@@ -678,7 +678,7 @@
 
   data_type = 1; //  for surveys
   echoinput << "Units:  0=numbers; 1=biomass; 2=F; >=30 for special patterns" << endl;
-  echoinput << "Errtype:  -1=normal; 0=lognormal; >0=T" << endl;
+  echoinput << "Errtype:  -2=gamma(future); -1=normal; 0=lognormal ; 1=lognormal w/ biascorr; >1XXX=T-dist with DF=XXX" << endl;
```
Comment (Cole-Monnahan-NOAA) [inline SS_readdata_330.tpl:681]: df=1 is a cauchy which probably we never want to use, so why not let Errtype>1 be t-dist with df=Errtype?

Comment (Cole-Monnahan-NOAA) [reply SS_readdata_330.tpl:681]: Actually a potentially easier implementation to leave the t-dist alone is to just use -2 for lognormal bias corr and I'll use -3 for gengamma later. Then you don't have to change anything related to t-dist, I think...

**Review by Cole-Monnahan-NOAA — APPROVED** [review]
The equations check out, although I must admit I did get a bit lost in the float options. But I feel confident the lognormal with bias correction and Q as a parameter is correct.

My minor comments are mostly about how this may break backward compatibility with t-distributions in use already. 

Edit: I should add that I was also convinced by the performance in the test case in [this comment](https://github.com/nmfs-ost/ss3-source-code/issues/535#issuecomment-1865380380).

**Comment by Rick-Methot-NOAA** [comment]
Thanks Cole for the good suggestions.  I originally was going to retain T-dist exactly as you suggest.  The reason I went with the distribution you reviewed is that it allowed for options >=0 to use log(surveyobs) and all options <0 would be in normal space.  I have never seen anyone use the T-dist (and wish we had a database of all applications so I would know!) , so I felt that the risk was low.  But the risk is not zero, so I will consider you recommendation.

**Comment by iantaylor-NOAA** [comment]
I found an email from 2020 in which Max Cardinale (@akatan999) referenced using the t-distribution for index likelihoods. It came up because the index plots in r4ss were not working correctly for t-distributed indices. I think the bug was present for at least 6 months and since nobody else reported the issue during that time, it's possible that nobody else was using the t-distribution feature.

**Comment by Rick-Methot-NOAA** [comment]
Thanks Ian.  I will maintain backward compatibility.

File: SS_objfunc.tpl (hunk 3 above)
Comment (Rick-Methot-NOAA) [reply SS_objfunc.tpl:139]: I will change such that

**Comment by kellijohnson-NOAA** [comment]
We are attempting to use the t-distribution for the age-1 index of Pacific Hake abundance in the 2024 stock assessment per your recommendations @Rick-Methot-NOAA when I talked to you last year about trying to fit estimates of zero.

**Comment by kellijohnson-NOAA** [comment]
Also, regarding my review of this code. I am right in the middle of an assessment so if you need my review it will have to wait until at least next week. Sorry.

**Review by Cole-Monnahan-NOAA — APPROVED** [review]

**Review by Cole-Monnahan-NOAA — APPROVED** [review]

**Pull request merged**